core/src/main/java/com/sequenceiq/cloudbreak/service/multiaz/MultiAzCalculatorService.java

```diff
@@ -0,0 +1,124 @@
+                    subnetAzPairs.put(value.getName(), value.getAvailabilityZone());
+                }
+                if (!Strings.isNullOrEmpty(value.getId())) {
+                    subnetAzPairs.put(value.getId(), value.getAvailabilityZone());
+                }
+            }
+        }
+        return subnetAzPairs;
+    }
+
+    public void calculateByRoundRobin(Map<String, String> subnetAzPairs, InstanceGroup instanceGroup) {
+        Map<String, Integer> subnetUsage = new HashMap<>();
+        Set<String> subnetIds = collectSubnetIds(instanceGroup);
+        initializeSubnetUsage(subnetAzPairs, subnetIds, subnetUsage);
+        collectCurrentSubnetUsage(instanceGroup, subnetUsage);
+
+        if (!subnetIds.isEmpty() && multiAzValidator.supportedForInstanceMetadataGeneration(instanceGroup)) {
+            for (InstanceMetaData instanceMetaData : instanceGroup.getAllInstanceMetaData()) {
+                if (Strings.isNullOrEmpty(instanceMetaData.getSubnetId())) {
+                    Integer numberOfInstanceInASubnet = searchTheSmallestInstanceCountForSubnets(subnetUsage);
+                    String leastUsedSubnetId = searchTheSmallestUsedSubnetID(subnetUsage, numberOfInstanceInASubnet);
+
+                    instanceMetaData.setSubnetId(leastUsedSubnetId);
+                    instanceMetaData.setAvailabilityZone(subnetAzPairs.get(leastUsedSubnetId));
+
+                    subnetUsage.put(leastUsedSubnetId, numberOfInstanceInASubnet + 1);
+                }
+            }
+        }
+    }
+
+    private Integer searchTheSmallestInstanceCountForSubnets(Map<String, Integer> subnetUsage) {
+        return Collections.min(subnetUsage.values());
+    }
+
+    private String searchTheSmallestUsedSubnetID(Map<String, Integer> subnetUsage, Integer currentNumber) {
+        return subnetUsage.entrySet()
+                .stream()
+                .filter(e -> e.getValue().equals(currentNumber))
+                .findFirst()
+                .get()
+                .getKey();
+    }
+
+    private void initializeSubnetUsage(Map<String, String> subnetAzPairs, Set<String> subnetIds, Map<String, Integer> subnetUsage) {
+        for (String key : subnetAzPairs.keySet()) {
+            if (subnetIds.contains(key)) {
+                subnetUsage.put(key, 0);
+            }
+        }
+    }
+
+    private void collectCurrentSubnetUsage(InstanceGroup instanceGroup, Map<String, Integer> subnetUsage) {
+
+        for (InstanceMetaData instanceMetaData : instanceGroup.getAllInstanceMetaData()) {
+            String subnetId = instanceMetaData.getSubnetId();
+            if (!Strings.isNullOrEmpty(subnetId)) {
+                Integer countOfInstances = subnetUsage.get(subnetId);
+                if (countOfInstances != null) {
+                    subnetUsage.put(subnetId, countOfInstances++);
+                } else {
+                    LOGGER.warn("Subnet id {} is not presented in the environment networks.", subnetId);
```

Comment:
Nits:
1. id -> ID
2. presented -> present
3. Perhaps print the value of `subnetUsage.keySet()` after `networks`